{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1227,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import string\n",
    "from tqdm import tqdm_notebook\n",
    "from tqdm import tqdm\n",
    "import datetime as dt\n",
    "import pickle\n",
    "from collections import Counter\n",
    "\n",
    "import nltk\n",
    "\n",
    "import pdaactconn as pc\n",
    "from trialexplorer.mesh_terms import MeSHCatalog\n",
    "from trialexplorer import AACTStudySet\n",
    "from trialexplorer import studysimilarity as ssim\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "255092 studies loaded!\n"
     ]
    }
   ],
   "source": [
    "# selecting all interventional studies\n",
    "conn = pc.AACTConnection(source=pc.AACTConnection.REMOTE)\n",
    "ss = AACTStudySet.AACTStudySet(conn= conn, tqdm_handler=tqdm_notebook)\n",
    "ss.add_constraint(\"study_type = 'Interventional'\")\n",
    "ss.load_studies()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfuly added these 1 dimensions: ['conditions']\n",
      "Failed to add these 0 dimensions: []\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2d4af51486634742bab1740a6f52eab1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=511), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Syncing the temp table temp_cur_studies in 511 chunks x 500 records each\n",
      "\n",
      "Creating index on the temp table\n",
      " - Loading dimension conditions\n",
      " -- Loading raw data\n",
      " -- Sorting index\n"
     ]
    }
   ],
   "source": [
    "# loading all dimensional data\n",
    "ss.add_dimensions('conditions')\n",
    "ss.refresh_dim_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>downcase_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nct_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>NCT00000102</td>\n",
       "      <td>10058348</td>\n",
       "      <td>Congenital Adrenal Hyperplasia</td>\n",
       "      <td>congenital adrenal hyperplasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NCT00000106</td>\n",
       "      <td>10057277</td>\n",
       "      <td>Rheumatic Diseases</td>\n",
       "      <td>rheumatic diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NCT00000108</td>\n",
       "      <td>10057274</td>\n",
       "      <td>Cardiovascular Diseases</td>\n",
       "      <td>cardiovascular diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NCT00000108</td>\n",
       "      <td>10057275</td>\n",
       "      <td>Coronary Disease</td>\n",
       "      <td>coronary disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NCT00000110</td>\n",
       "      <td>10057273</td>\n",
       "      <td>Obesity</td>\n",
       "      <td>obesity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id                            name  \\\n",
       "nct_id                                                  \n",
       "NCT00000102  10058348  Congenital Adrenal Hyperplasia   \n",
       "NCT00000106  10057277              Rheumatic Diseases   \n",
       "NCT00000108  10057274         Cardiovascular Diseases   \n",
       "NCT00000108  10057275                Coronary Disease   \n",
       "NCT00000110  10057273                         Obesity   \n",
       "\n",
       "                              downcase_name  \n",
       "nct_id                                       \n",
       "NCT00000102  congenital adrenal hyperplasia  \n",
       "NCT00000106              rheumatic diseases  \n",
       "NCT00000108         cardiovascular diseases  \n",
       "NCT00000108                coronary disease  \n",
       "NCT00000110                         obesity  "
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c = ss.dimensions['conditions']\n",
    "c.data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Extracting Adj and Verbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 420244/420244 [01:44<00:00, 4023.31it/s]\n"
     ]
    }
   ],
   "source": [
    "all_tag_counts = {}\n",
    "\n",
    "for cur_cond in tqdm(c.data['downcase_name'].values):\n",
    "    cur_tokens = nltk.word_tokenize(cur_cond)\n",
    "    pos_tags = nltk.pos_tag(cur_tokens)\n",
    "    for word, tag in pos_tags:\n",
    "        if tag not in all_tag_counts:\n",
    "            all_tag_counts[tag] = Counter()\n",
    "        all_tag_counts[tag][word] += 1\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['JJ', 'NN', 'NNS', ',', 'VBN', 'IN', 'VBP', 'RB', 'VBG', 'CC', 'POS', 'VB', 'DT', '(', ')', 'CD', 'VBZ', 'VBD', 'PRP$', 'FW', ':', 'TO', 'RP', 'SYM', 'UH', 'JJS', 'JJR', 'MD', 'NNP', \"''\", '.', '``', 'EX', 'WDT', '$', '#', 'WP', 'RBR', 'PRP', 'RBS', 'WRB', 'PDT', 'LS'])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_tag_counts.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JJ\n",
      "[('healthy', 10077), ('chronic', 9485), ('acute', 5392), ('coronary', 3841), ('pulmonary', 3835), ('metastatic', 3702), ('multiple', 3636), ('solid', 3468), ('renal', 3340), ('ovarian', 3291)]\n",
      "\n",
      "NN\n",
      "[('cancer', 39826), ('disease', 18944), ('stage', 15822), ('cell', 13729), ('lymphoma', 12040), ('carcinoma', 11373), ('breast', 10292), ('pain', 9798), ('leukemia', 9651), ('disorder', 9380)]\n",
      "\n",
      "NNS\n",
      "[('diabetes', 10498), ('diseases', 5203), ('disorders', 4874), ('infections', 4816), ('neoplasms', 4726), ('mellitus', 2812), ('tumors', 2599), ('volunteers', 1649), ('syndromes', 1332), ('patients', 1242)]\n",
      "\n",
      ",\n",
      "[(',', 28260)]\n",
      "\n",
      "VBN\n",
      "[('associated', 368), ('relapsed', 336), ('generalized', 271), ('related', 261), ('acquired', 247), ('transmitted', 208), ('treated', 201), ('delayed', 136), ('localized', 121), ('differentiated', 110)]\n",
      "\n",
      "IN\n",
      "[('of', 11445), ('with', 3751), ('in', 2915), ('for', 658), ('on', 460), ('after', 352), ('by', 312), ('without', 293), ('diffuse', 253), ('during', 240)]\n",
      "\n",
      "VBP\n",
      "[('mellitus', 3489), ('pectoris', 157), ('adult', 133), ('type', 106), ('virus', 75), ('simplex', 73), ('syndrome', 67), ('tendon', 44), ('cell', 42), ('chronic', 36)]\n",
      "\n",
      "RB\n",
      "[('back', 1413), ('non', 718), ('locally', 430), ('iv', 318), ('iii', 255), ('acute', 234), ('sexually', 210), ('fatty', 192), ('esophageal', 185), ('not', 174)]\n",
      "\n",
      "VBG\n",
      "[('smoking', 687), ('aging', 682), ('eating', 327), ('relapsing', 218), ('ewing', 184), ('ankylosing', 179), ('vomiting', 158), ('following', 151), ('requiring', 117), ('parenting', 111)]\n",
      "\n",
      "CC\n",
      "[('and', 7995), ('or', 1658), ('&', 94), ('q22', 31), ('excess', 19), ('q11.2', 18), ('plus', 14), ('q22.1', 12), ('q23.3', 12), ('newborn', 11)]\n",
      "\n",
      "POS\n",
      "[(\"'s\", 4080), (\"'\", 77), (\"'spine\", 1), (\"'in\", 1)]\n",
      "\n",
      "VB\n",
      "[('diffuse', 733), ('hemophilia', 176), ('severe', 128), ('ill', 95), ('moderate', 77), ('iv', 74), ('iii', 59), ('ii', 59), ('haemophilia', 58), ('mild', 50)]\n",
      "\n",
      "DT\n",
      "[('the', 4626), ('a', 828), ('tooth', 259), ('all', 188), ('no', 154), ('nonsmall', 72), ('an', 64), ('any', 57), ('nasal', 53), ('both', 26)]\n",
      "\n",
      "(\n",
      "[('(', 8597)]\n",
      "\n",
      ")\n",
      "[(')', 8586)]\n",
      "\n",
      "CD\n",
      "[('2', 5493), ('1', 2563), ('16', 513), ('3', 449), ('8', 178), ('21', 173), ('11q23', 148), ('5q', 114), ('15', 111), ('17', 106)]\n",
      "\n",
      "VBZ\n",
      "[('diabetes', 203), ('legs', 128), ('is', 86), ('lung', 82), ('lens', 61), ('acute', 40), (\"'s\", 27), ('adenoid', 20), ('hemophilia', 19), ('dust', 14)]\n",
      "\n",
      "VBD\n",
      "[('grade', 1224), ('induced', 299), ('related', 277), ('cleaved', 243), ('advanced', 236), ('left', 223), ('weight', 139), ('acquired', 124), ('glucose', 109), ('undifferentiated', 84)]\n",
      "\n",
      "PRP$\n",
      "[('its', 65), ('their', 19), ('her', 4), ('his', 4)]\n",
      "\n",
      "FW\n",
      "[('novo', 126), ('mellitus', 27), ('de', 22), ('vs.', 18), ('zika', 17), ('la', 7), ('klatskin', 7), ('knee', 7), ('et', 5), ('tourette', 5)]\n",
      "\n",
      ":\n",
      "[(';', 3054), ('-', 566), (':', 240), ('--', 50)]\n",
      "\n",
      "TO\n",
      "[('to', 1689)]\n",
      "\n",
      "RP\n",
      "[('back', 41), ('off', 9), ('down', 8), ('up', 8), ('out', 3), ('anesthesia', 1), ('anemia', 1), ('anterior', 1), ('breakdown', 1), ('teeth', 1)]\n",
      "\n",
      "SYM\n",
      "[('b', 84), ('stemi', 17), ('acl', 13), ('ibs', 12), ('chd', 8), ('dermatology/skin', 8), ('mrsa', 5), ('nafld', 5), ('sci', 5), ('bpd', 5)]\n",
      "\n",
      "UH\n",
      "[('brca1', 4), ('oh', 4), ('nsclc', 1), ('hmsh2', 1), ('e.g.', 1), ('lymphoma', 1), ('head/neck', 1)]\n",
      "\n",
      "JJS\n",
      "[('chest', 43), ('west', 22), ('least', 22), ('best', 3), ('most', 1), ('highest', 1), ('harvest', 1)]\n",
      "\n",
      "JJR\n",
      "[('lower', 458), ('older', 115), ('more', 76), ('less', 34), ('greater', 33), ('higher', 8), ('longer', 5), ('younger', 4), ('tangier', 4), ('shoulder', 3)]\n",
      "\n",
      "MD\n",
      "[('will', 30), ('onset', 18), ('optica', 17), ('may', 10), ('ca', 10), ('can', 8), ('must', 4), ('should', 2), ('could', 2), ('overweight', 2)]\n",
      "\n",
      "NNP\n",
      "[('/', 41), ('[', 32), ('+', 19), ('x', 11), ('<', 10), (']', 7), ('>', 5), ('=', 3), ('-cell', 3), ('xp11.2', 2)]\n",
      "\n",
      "''\n",
      "[(\"'\", 35), (\"''\", 22), ('’', 1), ('/', 1)]\n",
      "\n",
      ".\n",
      "[('.', 415)]\n",
      "\n",
      "``\n",
      "[('``', 22), ('`', 12)]\n",
      "\n",
      "EX\n",
      "[('meniere', 25), ('there', 4)]\n",
      "\n",
      "WDT\n",
      "[('that', 32), ('which', 14), ('khat', 2)]\n",
      "\n",
      "$\n",
      "[('>', 6), ('<', 3), ('≥', 2), ('[', 2), ('/', 1), ('=', 1), ('mellitus', 1)]\n",
      "\n",
      "#\n",
      "[('#', 3)]\n",
      "\n",
      "WP\n",
      "[('who', 137), ('whom', 5), ('hat', 3)]\n",
      "\n",
      "RBR\n",
      "[('lower', 47), ('less', 5), ('more', 3), ('laser', 2), ('better', 1), ('locus', 1), ('earlier', 1), ('lung', 1)]\n",
      "\n",
      "PRP\n",
      "[('we', 8), ('it', 7), ('self', 4), ('me', 3), ('s', 2), ('he', 2), ('them', 2), ('oabs', 1), ('oli', 1), ('us', 1)]\n",
      "\n",
      "RBS\n",
      "[('mild', 14), ('most', 2), ('best', 2), ('marrow', 1), ('mesothelin', 1), ('mellitus', 1), ('met', 1), ('yeast', 1)]\n",
      "\n",
      "WRB\n",
      "[('when', 9), ('how', 3), ('where', 2), ('severe', 1), ('meniere', 1)]\n",
      "\n",
      "PDT\n",
      "[('hemophilia', 3), ('haemophilia', 1)]\n",
      "\n",
      "LS\n",
      "[('tnbc', 6), ('oag', 2), ('tramadol', 1)]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for k, v in all_tag_counts.items():\n",
    "    print(k)\n",
    "    print(v.most_common(10))\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique adjective count : 6932\n",
      "Unique verb count : 2356\n"
     ]
    }
   ],
   "source": [
    "jj_count = 0\n",
    "vb_count = 0\n",
    "for k, v in all_tag_counts.items():\n",
    "    if k[:2] == 'JJ':\n",
    "        jj_count += len(v)\n",
    "    if k[:2] == 'VB':\n",
    "        vb_count += len(v)\n",
    "        \n",
    "print('Unique adjective count : %s' % jj_count)\n",
    "print('Unique verb count : %s' % vb_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_adj_and_vb(condition_str):\n",
    "    cur_cond = condition_str.lower()\n",
    "    cur_tokens = nltk.word_tokenize(cur_cond)\n",
    "    pos_tags = nltk.pos_tag(cur_tokens)\n",
    "    \n",
    "    all_jj, all_vb = [], []\n",
    "    for word, tag in pos_tags:\n",
    "        if tag[:2] == 'JJ':\n",
    "            if word not in all_jj:\n",
    "                all_jj.append(word)\n",
    "        if tag[:2] == 'VB':\n",
    "            if word not in all_vb:\n",
    "                all_vb.append(word)\n",
    "    return all_jj, all_vb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['contiguous', 'mixed'], [])"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extract_adj_and_vb('Stage II Contiguous Adult Diffuse Mixed Cell Lymphoma')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract type, grade, stage, AJCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 887,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_keywords = ['type', 'genotyp_', 'grade', 'stage', 'ajcc', 'hepatitis']\n",
    "\n",
    "word_nums = [\n",
    "    'zero', 'one', 'two', 'three', 'four',\n",
    "    'five', 'six', 'seven', 'eight', 'nine',\n",
    "    'ten', 'eleven', 'twelve', 'thirteen', 'fourteen',\n",
    "    'fifteen', 'sixteen', 'seventeen', 'eighteen', 'ninteen',\n",
    "    'twenty'\n",
    "]\n",
    "\n",
    "word2num = dict(zip(word_nums, range(0, 21)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### first just capture what comes after the space, see what our universe of data looks like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 888,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 420244/420244 [00:00<00:00, 472825.00it/s]\n"
     ]
    }
   ],
   "source": [
    "# distinct counters and dicts for each kw\n",
    "dict_list = []\n",
    "counter_list = []\n",
    "for i in range(len(all_keywords)):\n",
    "    dict_list.append({})\n",
    "    counter_list.append(Counter())\n",
    "    \n",
    "raw_counters = dict(zip(all_keywords, counter_list))\n",
    "orig_cond_dict = dict(zip(all_keywords, dict_list))\n",
    "\n",
    "for orig_cond in tqdm(c.data['downcase_name'].values):\n",
    "    for kw in all_keywords:\n",
    "        kw_splits = orig_cond.split(kw)\n",
    "        i = 0\n",
    "        for kw_split in kw_splits:\n",
    "            if i > 0:  # ignore first split\n",
    "                el_list = kw_split.strip().split(' ')\n",
    "                if len(el_list) > 0:\n",
    "                    el1 = el_list[0]\n",
    "                    raw_counters[kw][el1] += 1\n",
    "                                    \n",
    "                    if el1 in orig_cond_dict[kw]:\n",
    "                        orig_cond_dict[kw][el1].append(orig_cond)\n",
    "                    else:\n",
    "                        orig_cond_dict[kw][el1] = [orig_cond]\n",
    "            i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 889,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('iv', 3548),\n",
       " ('iii', 2599),\n",
       " ('ii', 1602),\n",
       " ('iiib', 993),\n",
       " ('i', 986),\n",
       " ('iiia', 898),\n",
       " ('iva', 740),\n",
       " ('iiic', 699),\n",
       " ('ivb', 685),\n",
       " ('renal', 655),\n",
       " ('iib', 491),\n",
       " ('iia', 458),\n",
       " ('ib', 301),\n",
       " ('ivc', 267),\n",
       " ('ia', 250),\n",
       " ('iic', 109),\n",
       " ('0', 91),\n",
       " ('small', 76),\n",
       " ('', 72),\n",
       " ('breast', 58)]"
      ]
     },
     "execution_count": 889,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counters['stage'].most_common(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 890,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['subjects with resectable local or locally advanced, non-metastatic (t2-t4, n0-n3, m0; stages ii and iii) and histologically-confirmed intestinal gc',\n",
       " 'cancer of the gi system-stages ii iii and iv',\n",
       " 'stages ii-iii breast cancer',\n",
       " 'chronic kidney disease stages 3-5',\n",
       " 'locoregional metastases in malignant melanoma stages iiib/c',\n",
       " 'stages iii skin melanoma',\n",
       " 'stages ii skin melanoma',\n",
       " 'breast cancer stages i through iii',\n",
       " 'figo stages ii to iv']"
      ]
     },
     "execution_count": 890,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig_cond_dict['stage']['s']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocessing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1023,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################\n",
    "# stage 1 preprocessing that occurs on the raw string:\n",
    "######################################################\n",
    "\n",
    "def convert_word2num(cur_text):\n",
    "    \"\"\" finds all of the english numbers in a string and replace with numerical numbers \"\"\"\n",
    "    rex = '\\s' + \"|\".join(word_nums) + '\\s'\n",
    "    found_list = re.findall(rex, cur_text)\n",
    "    if len(found_list) > 0:\n",
    "        for found_text in found_list:\n",
    "            sub_to = word2num[found_text.strip()]\n",
    "            cur_text = re.sub(rex, str(sub_to), cur_text)\n",
    "    return cur_text\n",
    "\n",
    "def convert_roman_to_str(cur_text):\n",
    "    \"\"\" remove roman numeral charsets \"\"\"\n",
    "    # only first 7\n",
    "    rn_map = {\n",
    "        chr(8544): 'i',\n",
    "        chr(8545): 'ii',\n",
    "        chr(8546): 'iii',\n",
    "        chr(8547): 'iv',\n",
    "        chr(8548): 'v',\n",
    "        chr(8549): 'vi',\n",
    "        chr(8550): 'vii',\n",
    "        chr(8560): 'i',\n",
    "        chr(8561): 'ii',\n",
    "        chr(8562): 'iii',\n",
    "        chr(8563): 'iv',\n",
    "        chr(8564): 'v',\n",
    "        chr(8565): 'vi',\n",
    "        chr(8566): 'vii',\n",
    "    }\n",
    "    \n",
    "    for k, v in rn_map.items():\n",
    "        cur_text = cur_text.replace(k, v)\n",
    "    return cur_text\n",
    "\n",
    "\n",
    "def preprocess_cond_str(cur_cond):\n",
    "    \"\"\" all preprocessing routines \"\"\"\n",
    "    # word2num\n",
    "    cur_cond = convert_word2num(cur_cond)\n",
    "    \n",
    "    # \"through\" delim\n",
    "    cur_cond = cur_cond.replace(' through ', '-')\n",
    "    \n",
    "    # \"and\" delim\n",
    "    cur_cond = cur_cond.replace(' and ', ',')\n",
    "    \n",
    "    # plural keywords\n",
    "    pl_kw = {\n",
    "        'stages': 'stage',\n",
    "        'grades': 'grade',\n",
    "        'types': 'type',\n",
    "        'genotypes': 'genotyp_',\n",
    "        'genotype': 'genotyp_'\n",
    "    }\n",
    "    for k, v in pl_kw.items():\n",
    "        cur_cond = cur_cond.replace(k, v)\n",
    "    \n",
    "    # connector delims\n",
    "    range_delims = {\n",
    "        \" - \": \"-\",\n",
    "        \" -\": \"-\",\n",
    "        \"- \": \"-\",\n",
    "        \" / \": \"/\",\n",
    "        \" /\": \"/\",\n",
    "        \"/ \": \"/\",\n",
    "    }\n",
    "    for k, v in range_delims.items():\n",
    "        cur_cond = cur_cond.replace(k, v)\n",
    "        \n",
    "    # inherient roman numerals in symbol\n",
    "    cur_cond = convert_roman_to_str(cur_cond)\n",
    "    \n",
    "    # ajcc abbrev.\n",
    "    ajcc_re = r'\\s?\\(american joint committee on cancer\\s?\\)\\s?' \n",
    "    cur_cond = re.sub(ajcc_re, '', cur_cond)\n",
    "    return cur_cond\n",
    "\n",
    "#########################################################\n",
    "# stage 2 preprocessing that occurs at the kw_split level\n",
    "#########################################################\n",
    "\n",
    "def preprocess_kw_split(kw_split):\n",
    "    \"\"\" preprocessing for the kw split level \"\"\"\n",
    "    \n",
    "    # replace leading dashes\n",
    "    kw_split = re.sub(r'^\\s?-', '', kw_split)\n",
    "    \n",
    "    # strip spaces\n",
    "    kw_split = kw_split.strip()\n",
    "    \n",
    "    return kw_split\n",
    "\n",
    "\n",
    "#########################################################\n",
    "# stage 3 preprocessing that occurs at the element level\n",
    "#########################################################\n",
    "\n",
    "def preprocess_element(element):\n",
    "    \"\"\" proprocessing for the element level \"\"\"\n",
    "    # replace roman numerals with numbers\n",
    "    # possible extension, extend to full number system\n",
    "    roman_map = {\n",
    "        'i': 1,\n",
    "        'ii': 2,\n",
    "        'iii': 3,\n",
    "        'iv': 4,\n",
    "        'v': 5,\n",
    "        'vi': 6,\n",
    "        'vii': 7,\n",
    "        'viii': 8,\n",
    "        'ix': 9,\n",
    "        'x': 10,\n",
    "        'xi': 11,\n",
    "        'xii': 12,\n",
    "        'xiii': 13,\n",
    "        'xiv': 14,\n",
    "        'xv': 15,\n",
    "        'xvi': 16,\n",
    "        'xvii': 17,\n",
    "        'xviii': 18,\n",
    "        'xix': 19,\n",
    "        'xx': 20,\n",
    "    }\n",
    "    \n",
    "    src_roman = '^i[vx]|^i{1,3}|^vi{1,3}|^xi[vx]|^xi{1,3}|^xvi{0,3}|^[xv]'\n",
    "    find_list = re.findall(src_roman, element)\n",
    "    if len(find_list) > 0:\n",
    "        found_str = find_list[0]\n",
    "        to_rep = roman_map[found_str]\n",
    "        element = re.sub(src_roman, str(to_rep), element)\n",
    "    \n",
    "    # replace brackets\n",
    "    element = element.replace(')', '')\n",
    "    element = element.replace('(', '')\n",
    "    \n",
    "    # replace 'diabetes' with ''\n",
    "    element = element.replace('diabetes', '')\n",
    "    \n",
    "    return element"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### extractor regex:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1142,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some kws use 1 regex pattern, ajcc uses another\n",
    "src1_delimiter = '[\\s*\\,\\-\\;\\)\\(]'\n",
    "\n",
    "src1 = r'^\\d[\\w\\(\\)]*' + src1_delimiter + '*|' + \\\n",
    "        '^i{2,3}[\\w\\(\\)]*' + src1_delimiter + '*|' + \\\n",
    "        '^vi{2,3}[\\w\\(\\)]*' + src1_delimiter + '*|' + \\\n",
    "        '^iv?[a-eg-mo-ru-z]?\\d+[\\w\\(\\)]*' + src1_delimiter + '+|' + \\\n",
    "        '^iv?[a-eg-mo-ru-z]?\\d+[\\w\\(\\)]*$|' + \\\n",
    "        '^iv?[a-eg-mo-ru-z]?\\d*' + src1_delimiter + '+|' + \\\n",
    "        '^iv?[a-eg-mo-ru-z]?\\d*$|' + \\\n",
    "        '^v\\d+[\\w\\(\\)]*' + src1_delimiter + '+|' + \\\n",
    "        '^v\\d+[\\w\\(\\)]*$|' + \\\n",
    "        '^v\\d*' + src1_delimiter + '+|' + \\\n",
    "        '^v\\d*$|' + \\\n",
    "        '^vi[a-z]\\d+[\\w\\(\\)]*' + src1_delimiter + '+|' + \\\n",
    "        '^vi[a-z]\\d+[\\w\\(\\)]*$|' + \\\n",
    "        '^vi[a-z]\\d*' + src1_delimiter + '+|' + \\\n",
    "        '^vi[a-z]\\d*$|' + \\\n",
    "        '^[A-Fa-f]\\d+[\\w\\(\\)]*' + src1_delimiter + '+|' + \\\n",
    "        '^[A-Fa-f]\\d+[\\w\\(\\)]*$|' + \\\n",
    "        '^[A-Fa-f]\\d*' + src1_delimiter + '+|' + \\\n",
    "        '^[A-Fa-f]\\d*$|' + \\\n",
    "        '^her\\d\\+?'\n",
    "\n",
    "#'^i[^A-za-z ][\\w\\(\\)]*[\\s*\\,\\-]*|' + \\\n",
    "#'^vi*[^A-za-z ][\\w\\(\\)]*[\\s*\\,\\-]*|' + \\\n",
    "\n",
    "src2 = r'^v\\d[\\w\\(\\)]*[\\s*\\,\\-]*'  # ajcc\n",
    "\n",
    "src3 = r'^[abcde]'\n",
    "regex_dict = {\n",
    "    'grade': src1,\n",
    "    'type': src1,\n",
    "    'genotyp_': src1,\n",
    "    'stage': src1,\n",
    "    'ajcc': src2,\n",
    "    'hepatitis': src3,\n",
    "}\n",
    "\n",
    "# pattern for delimiter\n",
    "src_delim = r'[\\s*\\,\\-]+$'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "melanoma stage iiib-ivm1a\n",
      "['melanoma ', ' iiib-ivm1a']\n",
      "iiib-ivm1a\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['3b', '4m1a'], ['-'])"
      ]
     },
     "execution_count": 1161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur_text = 'melanoma stage iiib-ivm1a'\n",
    "preped_cond = preprocess_cond_str(cur_text)\n",
    "print(preped_cond)\n",
    "extract_elements_delims(preped_cond, src1, 'stage', do_print=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1219,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_elements_delims(cur_text, regex_str, kw, do_element_pp=True, do_print=False):\n",
    "    elements = []\n",
    "    delimiters = []\n",
    "    \n",
    "    kw_splits = cur_text.split(kw)\n",
    "    if do_print:\n",
    "        print(kw_splits)\n",
    "        \n",
    "    if len(kw_splits) == 1:\n",
    "        return elements, delimiters  # kw not found\n",
    "    \n",
    "    for i, raw_kw_split in enumerate(kw_splits):\n",
    "        if i == 0:  # if the first split has no space, then the kw was part of another word\n",
    "            if len(raw_kw_split) != 0 and raw_kw_split[-1] in string.ascii_letters:\n",
    "                if do_print: print('not a real split')\n",
    "                break  # not a real split\n",
    "                \n",
    "        if i > 0:  # ignore first split\n",
    "            cur_kw_split = preprocess_kw_split(raw_kw_split)\n",
    "            if do_print: print(cur_kw_split)\n",
    "            do_loop = True\n",
    "            while do_loop:\n",
    "                # find the element and delimiter\n",
    "                cur_ele_list = re.findall(regex_str, cur_kw_split)\n",
    "                if len(cur_ele_list) > 0:\n",
    "                    remaining = cur_ele_list[0]\n",
    "                    # if we have an element then look for the delimiter\n",
    "                    delim = re.findall(src_delim, remaining)\n",
    "                    if len(delim) > 0:\n",
    "                        if delim[0].strip() == '-':\n",
    "                            delim_char = '-'\n",
    "                        else:\n",
    "                            delim_char = ','\n",
    "                        delimiters.append(delim_char)\n",
    "                        # remove delimiter\n",
    "                        remaining = re.sub(src_delim, '', remaining)\n",
    "                    else:\n",
    "                        do_loop = False\n",
    "\n",
    "                    # append element\n",
    "                    if do_element_pp:\n",
    "                        elements.append(preprocess_element(remaining).strip())\n",
    "                    else:\n",
    "                        elements.append(remaining.strip())\n",
    "                    # remove the element\n",
    "                    cur_kw_split = re.sub(regex_str, '', cur_kw_split)\n",
    "                else:\n",
    "                    do_loop = False\n",
    "    return elements, delimiters\n",
    "\n",
    "\n",
    "def digstr2parts(digstr):\n",
    "    \"\"\" returns a tuple of the digit and the string parts, if no num part, returns Null for that part \"\"\"\n",
    "    re_pattern = '^[0-9]+'\n",
    "    f_list = re.findall(re_pattern, digstr)\n",
    "    if len(f_list) == 0:\n",
    "        return None, digstr  # no number part\n",
    "    else:\n",
    "        num_part = int(f_list[0])\n",
    "        chr_part = re.sub(re_pattern, '', digstr)\n",
    "        return num_part, chr_part\n",
    "                \n",
    "\n",
    "def full_extract(orig_cond):\n",
    "    \"\"\" fully extract elements and delimiters from a condition string \"\"\"\n",
    "    preped_cond = preprocess_cond_str(orig_cond)\n",
    "    extracted_dict = {}\n",
    "    has_val = False\n",
    "    \n",
    "    for kw in all_keywords:\n",
    "        regex_str = regex_dict[kw]\n",
    "        do_element_pp = True\n",
    "        if kw == 'ajcc':\n",
    "            do_element_pp = False  # no not subsitute romans for ajcc\n",
    "        \n",
    "        elements, delimiters = extract_elements_delims(preped_cond, regex_str, kw, do_element_pp=do_element_pp)\n",
    "        if len(elements) > 0:\n",
    "            has_val = True\n",
    "            \n",
    "        \n",
    "        # count delimiters:\n",
    "        dash_delim_count = 0\n",
    "        for delim in delimiters:\n",
    "            if delim == '-':\n",
    "                dash_delim_count += 1\n",
    "        if dash_delim_count >= 1:\n",
    "            print(orig_cond)\n",
    "            print(elements, delimiters)\n",
    "        \n",
    "        # post process the elements and delimiters for all elements:\n",
    "        all_elements = post_process_element_delimiters(elements, delimiters)\n",
    "            \n",
    "        extracted_dict[kw] = all_elements\n",
    "        \n",
    "    if has_val:\n",
    "        return extracted_dict\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['3a', '3b', '3c', '4']"
      ]
     },
     "execution_count": 1220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# WIP\n",
    "def post_process_element_delimiters(elements, delimiters):\n",
    "    enumeratable_letters = 'abcdef'\n",
    "    \n",
    "    if '-' not in delimiters:\n",
    "        return elements\n",
    "    else:\n",
    "        cur_elements = []\n",
    "        prev_start_num = None\n",
    "        prev_end_num = None\n",
    "        prev_start_letter = None\n",
    "        prev_end_letter = None\n",
    "        \n",
    "        for i, delim in enumerate(delimiters):\n",
    "            # print(i, delim)\n",
    "            if delim == '-':\n",
    "                if len(elements) <= i + 1:\n",
    "                    # print('returning cond')\n",
    "                    return cur_elements  # return if there is no closing element for the current delimiter\n",
    "                else:\n",
    "                    start_num, start_letter = digstr2parts(elements[i])\n",
    "                    end_num, end_letter = digstr2parts(elements[i+1])\n",
    "                    #print(start_num, end_num, start_letter, end_letter)\n",
    "                    \n",
    "                    ####################\n",
    "                    # handle cases here\n",
    "                    ####################\n",
    "                    \n",
    "                    # case 1: we have both start and end nums\n",
    "                    if start_num and end_num:  # both not None\n",
    "                        first_loop = True\n",
    "                        for i in range(start_num, end_num):\n",
    "                            cur_ele = str(i)\n",
    "                            if first_loop: \n",
    "                                cur_ele += start_letter\n",
    "                            \n",
    "                            cur_elements.append(cur_ele)\n",
    "                            first_loop = False\n",
    "                            \n",
    "                        # on the end_num enumerate the letters\n",
    "                        if end_letter != '' and end_letter in enumeratable_letters:\n",
    "                            for j in range(ord('a'), ord(end_letter) + 1):\n",
    "                                append_chr = chr(j)\n",
    "                                cur_ele = str(end_num) + append_chr\n",
    "                                cur_elements.append(cur_ele)\n",
    "                        else:\n",
    "                            cur_elements.append(str(end_num))\n",
    "                    \n",
    "                    # case 2: have only letters - enumerate all the letters\n",
    "                    elif start_num is None and end_num is None:\n",
    "                        if start_letter != '' and end_letter != '' and \\\n",
    "                        start_letter in enumeratable_letters and \\\n",
    "                        end_letter in enumeratable_letters:\n",
    "                            for j in range(ord(start_letter), ord(end_letter) + 1):\n",
    "                                cur_ele = chr(j)\n",
    "                                cur_elements.append(cur_ele)\n",
    "                    \n",
    "                    # case 3: we have start num but no end nums - enumerate only the start num letters\n",
    "                    elif start_num and end_num is None:\n",
    "                        if start_letter != '' and end_letter != '' and \\\n",
    "                        start_letter in enumeratable_letters and \\\n",
    "                        end_letter in enumeratable_letters:\n",
    "                            for j in range(ord(start_letter), ord(end_letter) + 1):\n",
    "                                append_chr = chr(j)\n",
    "                                cur_ele = str(start_num) + append_chr\n",
    "                                cur_elements.append(cur_ele)\n",
    "                                \n",
    "                    # case 4: no start num, but end num and prev num - enumerate all numbers except first\n",
    "                    # and no end letter\n",
    "                    elif start_num is None and end_num and prev_start_num and end_letter == '':\n",
    "                        for i in range(prev_start_num + 1, end_num + 1):\n",
    "                            cur_ele = str(i)\n",
    "                            cur_elements.append(cur_ele)\n",
    "                            \n",
    "                    # case 5: no start num, but end num and prev num - enumerate all numbers except first\n",
    "                    # and has end letter\n",
    "                    elif start_num is None and end_num and prev_start_num and end_letter != '':\n",
    "                        for i in range(prev_start_num + 1, end_num):\n",
    "                            cur_ele = str(i)\n",
    "                            cur_elements.append(cur_ele)\n",
    "                            \n",
    "                        if end_letter in enumeratable_letters:\n",
    "                            for j in range(ord('a'), ord(end_letter) + 1):\n",
    "                                append_chr = chr(j)\n",
    "                                cur_ele = str(end_num) + append_chr\n",
    "                                cur_elements.append(cur_ele)\n",
    "                        else:\n",
    "                            cur_ele = str(end_num) + end_letter\n",
    "                            cur_elements.append(cur_ele)\n",
    "                        \n",
    "      \n",
    "                    # storing prev loop values\n",
    "                    prev_start_num = start_num\n",
    "                    prev_end_num = end_num\n",
    "                    prev_start_letter = start_letter\n",
    "                    prev_end_letter = end_letter\n",
    "                        \n",
    "                            \n",
    "        return cur_elements\n",
    "                \n",
    "#post_process_element_delimiters(['b'], ['-'])\n",
    "post_process_element_delimiters(['3a', 'c', '4'], ['-', '-', ','])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# testing full extract:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1221,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 32%|███▏      | 20785/64114 [00:00<00:00, 69076.26it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "early stage breast cancer (stage 1-3)\n",
      "['1', '3'] ['-']\n",
      "low grade b-cell lymphoma, not otherwise specified\n",
      "['b'] ['-']\n",
      "type i-allergy\n",
      "['1'] ['-']\n",
      "breast cancer early stage breast cancer (stage 1-3)\n",
      "['1', '3'] ['-']\n",
      "stages ii-iii breast cancer\n",
      "['2', '3'] ['-', ',']\n",
      "patients with a diagnosis of who grade i-iii meningioma and hemangiopericytoma with adequate bone marrow function\n",
      "['1', '3'] ['-', ',']\n",
      "we investigated the relationship between plasma fgf23 levels and endothelial dysfunction in a sizable series of incident stage 3-4 ckd patients.\n",
      "['3', '4'] ['-', ',']\n",
      "early stage breast cancer (stage 0-iii)\n",
      "['0', '3'] ['-']\n",
      "stage i-iii small cell lung cancer\n",
      "['1', '3'] ['-', ',']\n",
      "stage i-iii non-small cell lung cancer\n",
      "['1', '3'] ['-', ',']\n",
      "patients with stage 2-4 chronic kidney disease\n",
      "['2', '4'] ['-', ',']\n",
      "indolent or intermediate grade b-cell malignancy\n",
      "['b'] ['-']\n",
      "chronic kidney disease stages 3-5\n",
      "['3', '5'] ['-']\n",
      "erbb2-positive stage i-iii breast cancer\n",
      "['1', '3'] ['-', ',']\n",
      "stage ii-iii\n",
      "['2', '3'] ['-']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 64%|██████▍   | 41035/64114 [00:00<00:00, 67954.20it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stage i-ivb colon cancer\n",
      "['1', '4b'] ['-', ',']\n",
      "stage i-ivb rectal cancer\n",
      "['1', '4b'] ['-', ',']\n",
      "high grade b-cell lymphoma with myc and bcl2 or bcl6 rearrangements\n",
      "['b'] ['-']\n",
      "type-2-diabetes mellitus\n",
      "['2'] ['-']\n",
      "bone pain in stage i - iii breast cancer\n",
      "['1', '3'] ['-', ',']\n",
      "inoperable esophageal cancer stage i-iii\n",
      "['1', '3'] ['-']\n",
      "high-grade b-cell lymphoma (hgbl)\n",
      "['b'] ['-']\n",
      "locally advanced or metastatic non small cell lung cancer stage iiib - iv\n",
      "['3b', '4'] ['-']\n",
      "stage iii-iv or recurrent endometrial cancer\n",
      "['3', '4'] ['-', ',']\n",
      "post-menopausal er+ stage i-iiia primary operable breast cancer\n",
      "['1', '3a'] ['-', ',']\n",
      "newly diagnosed operative invasive ductal breast carcinoma stage i-ii\n",
      "['1', '2'] ['-']\n",
      "type 2-diabetes\n",
      "['2'] ['-']\n",
      "stage ia-iia breast cancer\n",
      "['1a', '2a'] ['-', ',']\n",
      "figo stage iii-iv\n",
      "['3', '4'] ['-']\n",
      "high grade b-cell lymphoma\n",
      "['b'] ['-']\n",
      "refractory high grade b-cell lymphoma with myc and bcl2 or bcl6 rearrangements\n",
      "['b'] ['-']\n",
      "recurrent high grade b-cell lymphoma with myc and bcl2 or bcl6 rearrangements\n",
      "['b'] ['-']\n",
      "low-grade b-cell lymphoma\n",
      "['b'] ['-']\n",
      "breast cancer stages i through iii\n",
      "['1', '3'] ['-']\n",
      "localized stage i-iii esophageal cancer\n",
      "['1', '3'] ['-', ',']\n",
      "invasive breast cancer (stage i-iii)\n",
      "['1', '3'] ['-']\n",
      "stage i-iii pancreatic cancer\n",
      "['1', '3'] ['-', ',']\n",
      "stage i-iii colon cancer\n",
      "['1', '3'] ['-', ',']\n",
      "acute graft versus host disease grade ii-iv\n",
      "['2', '4'] ['-']\n",
      "locally advanced or metastatic non-small-cell lung cancer (stage iiib-iv)\n",
      "['3b', '4'] ['-']\n",
      "figo stage iic-iv ovarian carcinoma\n",
      "['2c', '4'] ['-', ',']\n",
      "high grade b-cell lymphoma with myc, bcl2, and bcl6 rearrangements\n",
      "['b'] ['-']\n",
      "chronic kidney disease stage 3-5\n",
      "['3', '5'] ['-']\n",
      "high grade b-cell lymphoma (hgbcl)\n",
      "['b'] ['-']\n",
      "ann arbor stage iii b-cell non-hodgkin lymphoma\n",
      "['3', 'b'] [',', '-']\n",
      "ann arbor stage iv b-cell non-hodgkin lymphoma\n",
      "['4', 'b'] [',', '-']\n",
      "ann arbor stage i b-cell non-hodgkin lymphoma\n",
      "['1', 'b'] [',', '-']\n",
      "ann arbor stage ii b-cell non-hodgkin lymphoma\n",
      "['2', 'b'] [',', '-']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 86%|████████▌ | 54846/64114 [00:00<00:00, 68471.48it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stage ib-iiia non-small cell lung carcinoma\n",
      "['1b', '3a'] ['-', ',']\n",
      "high-grade b-cell lymphoma with myc, bcl2, and bcl6 rearrangements\n",
      "['b'] ['-']\n",
      "gaucher disease type 1-gaucher disease type 3\n",
      "['1', '3'] ['-']\n",
      "stage 0-iii breast cancer\n",
      "['0', '3'] ['-', ',']\n",
      "stage iiib-c melanoma\n",
      "['3b', 'c'] ['-', ',']\n",
      "glycogen storage disease type ii-pompe's disease\n",
      "['2'] ['-']\n",
      "small intestinal high grade b-cell lymphoma, not otherwise specified\n",
      "['b'] ['-']\n",
      "high grade b-cell lymphoma, not otherwise specified\n",
      "['b'] ['-']\n",
      "unresectable stage iiib-iv malignant melanoma\n",
      "['3b', '4'] ['-', ',']\n",
      "high-grade b-cell lymphoma\n",
      "['b'] ['-']\n",
      "refractory high grade b-cell lymphoma\n",
      "['b'] ['-']\n",
      "recurrent high grade b-cell lymphoma\n",
      "['b'] ['-']\n",
      "ckd stage 1-4\n",
      "['1', '4'] ['-']\n",
      "stage i-ii head and neck cancer\n",
      "['1', '2'] ['-', ',']\n",
      "advanced (stage iiib-c-iv) ovarian, primary peritoneal and fallopian tube cancer\n",
      "['3b', 'c', '4'] ['-', '-', ',']\n",
      "prevention of hpv types 16- and 18-related cervical cancer, cervical intraepithelial neoplasia (cin) 1/2/3, and cervical adenocarcinoma in situ\n",
      "['16', '18'] [',', '-']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 64114/64114 [00:00<00:00, 68283.41it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "math teachers (grades 2-8)\n",
      "['2', '8'] ['-']\n",
      "math students (grades 2-8)\n",
      "['2', '8'] ['-']\n",
      "stage ii-ivb operable hnscc oral cavity\n",
      "['2', '4b'] ['-', ',']\n",
      "high grade b-cell lymphoma with myc and bcl2 and/or bcl6 rearrangements\n",
      "['b'] ['-']\n",
      "stage ia - ib non small cell lung cancer\n",
      "['1a', '1b'] ['-', ',']\n",
      "early stage non-small-cell lung cancer (stage 1-2)\n",
      "['1', '2'] ['-']\n",
      "melanoma stage iiib-ivm1a\n",
      "['3b', '4m1a'] ['-']\n",
      "stage ib-iib cutaneous t-cell lymphoma\n",
      "['1b', '2b'] ['-', ',']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "all_extracted = {}\n",
    "for orig_cond in tqdm(c.data['downcase_name'].unique()):\n",
    "    extracted_dict = full_extract(orig_cond)\n",
    "    if extracted_dict:\n",
    "        all_extracted[orig_cond] = extracted_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'type': [],\n",
       " 'genotyp_': [],\n",
       " 'grade': [],\n",
       " 'stage': ['1', '2', '3'],\n",
       " 'ajcc': [],\n",
       " 'hepatitis': []}"
      ]
     },
     "execution_count": 1226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_extracted['erbb2-positive stage i-iii breast cancer']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All texts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['type', 'genotyp_', 'grade', 'stage', 'ajcc', 'hepatitis']"
      ]
     },
     "execution_count": 1145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_keywords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1146,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 420244/420244 [00:05<00:00, 72694.10it/s]\n"
     ]
    }
   ],
   "source": [
    "# distinct counters and dicts for each kw\n",
    "dict_list = []\n",
    "counter_list = []\n",
    "skipped_list = []\n",
    "for i in range(len(all_keywords)):\n",
    "    dict_list.append({})\n",
    "    skipped_list.append([])\n",
    "    counter_list.append(Counter())\n",
    "    \n",
    "raw_counters = dict(zip(all_keywords, counter_list))\n",
    "orig_kw_dict = dict(zip(all_keywords, dict_list))\n",
    "skipped_dict = dict(zip(all_keywords, skipped_list))\n",
    "\n",
    "#for orig_cond in tqdm(['diphtheria; haemophilus influenzae type b; hepatitis b; tetanus; whole cell pertussis']):\n",
    "for orig_cond in tqdm(c.data['downcase_name'].values):\n",
    "    preped_cond = preprocess_cond_str(orig_cond)\n",
    "    for kw in all_keywords:\n",
    "        regex_str = regex_dict[kw]\n",
    "        do_element_pp = True\n",
    "        if kw == 'ajcc':\n",
    "            do_element_pp = False  # no not subsitute romans for ajcc\n",
    "        \n",
    "        elements, delimiters = extract_elements_delims(preped_cond, regex_str, kw, do_element_pp=do_element_pp)\n",
    "        \n",
    "        if len(elements) == 0 and kw in preped_cond:\n",
    "            if preped_cond not in skipped_dict[kw]:\n",
    "                skipped_dict[kw].append(preped_cond)\n",
    "        \n",
    "        for element in elements:\n",
    "            raw_counters[kw][element] += 1\n",
    "            \n",
    "            if element in orig_kw_dict[kw]:\n",
    "                orig_kw_dict[kw][element].append(orig_cond)\n",
    "            else:\n",
    "                orig_kw_dict[kw][element] = [orig_cond]\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['end-stage renal disease',\n",
       " 'small cell lung cancer ex10sive stage',\n",
       " 'end stage renal disease',\n",
       " 'ex10sive stage small cell lung cancer',\n",
       " 'limited stage small cell lung cancer',\n",
       " 'renal disease, end-stage',\n",
       " 'end-stage kidney disease',\n",
       " 'end stage liver disease',\n",
       " 'early-stage breast cancer',\n",
       " 'third stage of labour',\n",
       " 'early stage esophageal adenocarcinoma',\n",
       " 'end-stage renal failure',\n",
       " 'stage, intraocular melanoma',\n",
       " 'ex10sive stage unresectable',\n",
       " \"early stage parkinson's disease\",\n",
       " 'renal failure, end stage',\n",
       " 'end-stage renal disease patients on hemodialysis',\n",
       " 'early stage breast cancer',\n",
       " \"advanced stage diffuse large b-cell non-hodgkin's lymphoma\",\n",
       " 'early stage parkinson disease',\n",
       " 'end-stage chronic renal disease',\n",
       " 'end stage renal disease (esrd)',\n",
       " \"advanced stage parkinson's disease\",\n",
       " 'labor stage, first',\n",
       " 'stage t3-4nxm0 gastric cancer',\n",
       " 'ex10sive stage lung small cell carcinoma',\n",
       " 'end-stage heart failure',\n",
       " 'endstage renal disease',\n",
       " 'second stage labor',\n",
       " 'end stage macular degeneration',\n",
       " 'end stage renal failure, hemodialysis',\n",
       " 'renal disease, end stage',\n",
       " 'inoperable stage i/ii non-small cell lung cancer',\n",
       " \"early-stage parkinson's disease\",\n",
       " 'ex10sive-stage small cell lung cancer',\n",
       " 'ex10sive stage small cel lung cancer',\n",
       " 'end stage kidney disease',\n",
       " 'second stage of labor',\n",
       " 'solid tumor regardless of localization,stage',\n",
       " 'end stage cardiac failure',\n",
       " 'cleavage stage, ovum',\n",
       " 'early stage endometrial cancer',\n",
       " 'end stage renal failure on dialysis',\n",
       " 'end stage liver failure',\n",
       " 'third stage of labor',\n",
       " 'ex10sive stage lung cancer',\n",
       " 'ex10sive stage small cell lung carcinoma',\n",
       " 'end-stage liver disease',\n",
       " 'cancer (advanced stage)',\n",
       " 'end stage renal failure',\n",
       " 'early stage hepatocellular carcinoma',\n",
       " 'early stage non-small cell lung cancer',\n",
       " 'inoperable early stage non-small cell lung cancer',\n",
       " 'end-stage renal disease (esrd)',\n",
       " 'clinical stage t2b or less of prostate cancer',\n",
       " 'labor stage, second',\n",
       " 'end stage renal failure with renal transplant',\n",
       " 'solid tumor in advanced stage',\n",
       " 'pediatric end stage renal disease',\n",
       " 'perit1al serous-type advanced stage',\n",
       " 'end-stage disease',\n",
       " '1-stage bilateral total hip replacement',\n",
       " 'stage, breast cancer',\n",
       " 'labor, second stage',\n",
       " 'labor, first stage',\n",
       " 'end stage heart disease',\n",
       " 'early-stage squamous cell carcinoma of the oropharynx',\n",
       " 'early stage prostate cancer',\n",
       " 'anemia in end stage renal disease',\n",
       " 'prolonged first stage of labor',\n",
       " 'acute renal failure (as defined by the \"failure\" stage of the rifle classification)',\n",
       " 'patch/plaque stage mycosis fungoides',\n",
       " 'end stage renal disease/hemodialysis',\n",
       " 'cancer in remission (any type or stage)',\n",
       " 'advanced stage cancer',\n",
       " 'limited stage',\n",
       " 'small cell lung cancer limited stage',\n",
       " 'high model for end-stage liver disease (meld) score',\n",
       " 'her2-positive early stage breast cancer',\n",
       " 'end stage renal disease,kidney transplantation',\n",
       " 'non small cell lung cancer stage',\n",
       " 'end stage cancer',\n",
       " 'secons stage of labor',\n",
       " 'stage, prostate cancer',\n",
       " 'advanced stage non small cell lung cancer',\n",
       " 'robotic hysterectomy (benign indications/early-stage cancer)',\n",
       " 'early-stage glottic larynx cancer',\n",
       " 'ex10sive-stage small-cell lung cancer',\n",
       " 'anemia of end stage renal disease',\n",
       " 'early stage estrogen receptor (er) positive breast cancer',\n",
       " 'end-stage renal disease requiring haemodialysis',\n",
       " 'management of the second stage of labor in obese nulliparous women by either passive descent or immediate pushing.',\n",
       " 'advanced endstage solid carcinomas in adults',\n",
       " 'early stage breast carcinoma',\n",
       " 'early-stage breast carcinoma',\n",
       " 'end stage renal disease requiring hemodialysis',\n",
       " 'renal failure, end-stage',\n",
       " 'advanced stage solid tumors',\n",
       " 'stage, colon cancer',\n",
       " 'second stage of labour',\n",
       " 'prolonged second stage of labor',\n",
       " 'stage i/ii extranodal nk/t-cell lymphoma',\n",
       " 'ex10sive-stage small cell lung cancer (sclc)',\n",
       " \"late stage parkinson's disease\",\n",
       " 'advanced stage non-small cell lung cancer',\n",
       " 'cutaneous melanoma by ajcc v7 stage',\n",
       " 'limited stage small cell lung carcinoma',\n",
       " '2nd stage of labor',\n",
       " 'labor stage, third/drug effects',\n",
       " 'first labor stage',\n",
       " 'primary myelofibrosis, prefibrotic stage',\n",
       " 'stage ct2 n1-3 m0 or ct3-t4a n0 or n1-3 m0',\n",
       " 'renal insufficiency or renal failure &or end-stage renal disease',\n",
       " 'end stage renal diseases',\n",
       " 'labor stage, third',\n",
       " 'end-stage kidney failure',\n",
       " 'very-early stage',\n",
       " 'early stage',\n",
       " 'end-stage liver disease (esld)',\n",
       " 'end-stage chronic kidney failure',\n",
       " 'end stage chronic obstructive airways disease',\n",
       " 'end stage renal disease on hemodialysis',\n",
       " 'stage, pancreatic cancer',\n",
       " 'esrd (end-stage renal disease)',\n",
       " 'prolonged first stage labor-prevention',\n",
       " 'platinum refractory ex10sive-stage small cell lung carcinoma',\n",
       " 'kidney disease, end-stage',\n",
       " 'lung cancer, limited stage small cell',\n",
       " 'esophageal squamous cell carcinoma stage ct3-4an0-1m0',\n",
       " 'active second stage',\n",
       " 'endstage kidney disease',\n",
       " 'non-small cell lung cancer stage',\n",
       " 'small cell lung cancer, limited stage',\n",
       " 'primary myelofibrosis, fibrotic stage',\n",
       " 'inrg stage ms',\n",
       " 'inrg stage l2',\n",
       " 'inrg stage m',\n",
       " 'cervical cancer stage',\n",
       " 'small cell lung ex10sive stage',\n",
       " 'end-stage glaucoma',\n",
       " 'melanoma stage',\n",
       " 'end stage organ failure',\n",
       " 'heart disease end stage',\n",
       " 'end stage heart failure',\n",
       " 'pregnant women who requested epidural analgesia in the early stage of labor',\n",
       " 'chronic obstructive pulmonary disease end stage',\n",
       " 'anemia in end-stage renal disease',\n",
       " 'advanced or metastasis stage',\n",
       " 'pancreatic cancer stage',\n",
       " 'early stage hcc',\n",
       " 'end stage renal disease on dialysis (diagnosis)',\n",
       " 'hepatocellular carcinoma by bclc stage',\n",
       " 'breast cancer stage',\n",
       " 'primary open-angle glaucoma, mild stage',\n",
       " 'patients with end stage renal disease on hemodialysis',\n",
       " 'end stage disease',\n",
       " 'end stage renal disease on dialysis',\n",
       " 'third stage of delivery',\n",
       " 'advanced stage',\n",
       " 'intermediate stage of hepatocellular carcinoma',\n",
       " 'colon cancer stage',\n",
       " 'recurrent ex10sive stage small cell lung carcinoma',\n",
       " 'refractory ex10sive stage small cell lung carcinoma',\n",
       " 'prodromal stage',\n",
       " 'early-stage',\n",
       " 'limited stage lung small cell carcinoma',\n",
       " 'early-stage nasopharyngeal carcinoma',\n",
       " 'labor; prolonged, first stage',\n",
       " \"idiopathic parkinson's disease (at later stage)\",\n",
       " 'patients with advanced stage solid tumors',\n",
       " 'nonsmall cell lung cancer stage',\n",
       " 'esophageal squamous cell carcinoma by ajcc v8 stage',\n",
       " 'bladder cancer stage']"
      ]
     },
     "execution_count": 1147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "skipped_dict['stage']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1148,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('4', 3613),\n",
       " ('3', 2673),\n",
       " ('2', 1635),\n",
       " ('1', 1017),\n",
       " ('3b', 1012),\n",
       " ('3a', 905),\n",
       " ('4a', 741),\n",
       " ('3c', 702),\n",
       " ('4b', 689),\n",
       " ('2b', 494),\n",
       " ('2a', 461),\n",
       " ('1b', 305),\n",
       " ('4c', 267),\n",
       " ('1a', 252),\n",
       " ('2c', 110),\n",
       " ('0', 94),\n",
       " ('5', 33),\n",
       " ('1c', 29),\n",
       " ('b', 23),\n",
       " ('c', 11)]"
      ]
     },
     "execution_count": 1148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counters['stage'].most_common(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chronic-kidney disease stage 5d on stable hemodialysis',\n",
       " 'chronic-kidney disease stage 5d on stable hemodialysis']"
      ]
     },
     "execution_count": 1160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig_kw_dict['stage']['5d']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1149,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['adult nasal type extranodal nk/t-cell lymphoma',\n",
       " 'stereotyped behavior',\n",
       " 'adult type ovarian granulosa cell tumor',\n",
       " 'childhood nasal type extranodal nk/t-cell lymphoma',\n",
       " 'salivary gland malignant mixed cell type tumor',\n",
       " 'dementia, alzheimer type',\n",
       " 'neoplasms by histologic type',\n",
       " 'schizoaffective disorder, manic type',\n",
       " \"psychosis,behavioral disturbances associated with dementia of the alzheimer's type\",\n",
       " 'locally advanced undifferentiated carcinoma nasopharyngeal type ucnt',\n",
       " 'dementia of alzheimer type',\n",
       " 'pneumonic-type adenocarcinoma (p-adc)',\n",
       " 'senile dementia, alzheimer type',\n",
       " 'other subtype of b-cell lymphoma',\n",
       " 'hypersensitivity, delayed type',\n",
       " '10sion-type headache',\n",
       " 'wildtype',\n",
       " 'intraabdominal cancers (various type)',\n",
       " \"dementia of the alzheimer's type\",\n",
       " 'childhood botryoid-type embryonal rhabdomyosarcoma',\n",
       " 'influenza a virus, h5n1 subtype',\n",
       " 'high-type anal fistulae',\n",
       " 'psoriasis-type psoriasis',\n",
       " 'plaque-type psoriasis',\n",
       " '10sion type headache',\n",
       " 'adult type diabetes mellitus',\n",
       " 'to compare the effect of 2 type of iols',\n",
       " 'high blood b-type (or brain) natriuretic peptide (bnp) level',\n",
       " 't/null cell systemic type',\n",
       " 'mixed phenotype acute leukemia',\n",
       " 'ovarian sex cord-stromal tumor of mixed or unclassified cell type',\n",
       " 'influenza a subtype h5n1 infection',\n",
       " 'type of delivery',\n",
       " 'somatic complaints of multiple type',\n",
       " 'dehiscent type of alveolar b1 defect',\n",
       " 'type of general anesthesia',\n",
       " 'extranodal nk/t-cell lymphoma nasal type',\n",
       " 'enteropathy-type t-cell lymphoma',\n",
       " 'ischaemic type biliary lesion',\n",
       " 'cervical carcinoma, non-squamoustype',\n",
       " 'chronic plaque-type psoriasis',\n",
       " 'influenza a virus, h1n1 subtype',\n",
       " 'cytochrome p450 phenotype,genotyp_ metrics',\n",
       " 'schizophrenia, undifferentiated type',\n",
       " 'hip fracture of intertrochanteric type',\n",
       " 'malignant lymphoma, stem cell type',\n",
       " 'diffuse-type giant cell tumor',\n",
       " 'all type of infections',\n",
       " 'single bout of resistance type exercise',\n",
       " 'extranodal nk-t-cell lymphoma, nasal,nasal-type',\n",
       " 'alzheimer type dementia',\n",
       " 'gastric intestinal type adenocarcinoma',\n",
       " \"dementia alzheimer's type\",\n",
       " \"alzheimer's type dementia\",\n",
       " 'dementia of the alzheimer type',\n",
       " 'refractory mixed type melasma',\n",
       " 'autoimmune hemolytic disease (cold type) (warm type)',\n",
       " 'myelodysplastic syndromes (mds) other than ra or rars subtype',\n",
       " 'kras wild type',\n",
       " 'moderate to severe plaque-type psoriasis',\n",
       " 'ductal histologic type',\n",
       " 'kras wildtype',\n",
       " 'lung adenocarcinoma, mixed subtype',\n",
       " 'perit1al serous-type advanced stage',\n",
       " '10sional-type headache',\n",
       " 'extremely well differentiated intestinal-type adenocarcinoma',\n",
       " '10sion-type headache (episodic)',\n",
       " 'headache, 10sion-type',\n",
       " 'episodic 10sion type headache',\n",
       " 'nasal,nasal-type nk/t-cell lymphoma',\n",
       " 'kras wild type colorectal cancer',\n",
       " 'primary amyloidosis of light chain type',\n",
       " 'mature b-cell leukemia burkitt-type',\n",
       " 'influenza a subtype h7n7 infection',\n",
       " 'psoriasis, plaque-type psoriasis',\n",
       " 'locally advanced or metastatic non-clear cell type renal cell carcinoma',\n",
       " 'at10tion deficit/hyperactivity disorder combined type',\n",
       " 'adhd predominantly inat10tive type',\n",
       " 'adhd predominantly hyperactivity type',\n",
       " 'refractory primary-cutaneous large b-cell lymphoma (leg-type)',\n",
       " 'delayed sleep phase type circadian rhythm sleep disorder',\n",
       " 'blood, injection, injury type phobia',\n",
       " 'low-density-lipoprotein-type',\n",
       " 'acceptability of different contraceptive injection type',\n",
       " 'seizures (incl subtype)',\n",
       " 'wild type braf stage iv melanoma',\n",
       " 'moderate to severe chronic plaque-type psoriasis',\n",
       " 'metastatic cancer (different solid tumour type)',\n",
       " '10osynovial giant cell tumor localized or diffused type',\n",
       " 'plaque type psoriasis',\n",
       " 'haemophilus influenzae serotype b (hib)',\n",
       " 'rosacea subtype 1 (erythematotelangiectatic)',\n",
       " 'extranodal nk/t-cell lymphoma, nasal type',\n",
       " 'ras-wildtype colorectal cancer',\n",
       " 'cancer in remission (any type or stage)',\n",
       " 'schizophrenia; negative type',\n",
       " 'adhd-inat10tive type',\n",
       " 'adhd-combined type',\n",
       " 'carcinoma, diffuse type',\n",
       " 'intestinal type adenocarcinoma of stomach',\n",
       " 'egfr unknown or wild-type',\n",
       " 'solid tumor with p53 wild type status',\n",
       " 'episodic 10sion-type headache',\n",
       " 'brca1,brca2 mutation carrier,brcaness phenotype',\n",
       " 'adhesive remnant index for type of bond failure',\n",
       " 'nasal type extranodal nk/t-cell lymphoma',\n",
       " 'methylmalonic aciduria,homocystinuria,cblc type',\n",
       " 'type of seizures,epileptic syndromes',\n",
       " 'liver only metastasis from kras exon 2 wild type (under protocol 1.0-1.2 edition),ras wild type (under protocol 2.0 edition) colorectal cancer',\n",
       " 'influenza due to influenza a virus subtype h5n2',\n",
       " 'all type of cancers with bony metastasis',\n",
       " 'cd79 mutant or abc-subtype diffuse large b-cell lymphoma',\n",
       " \"alzheimer's type\",\n",
       " 'dementia of type alzheimer',\n",
       " 'psoriasis, plaque-type',\n",
       " 'haemophilus influenzae serotype b disease',\n",
       " 'anorexia nervosa restricting type',\n",
       " 'chronic 10sion-type headache',\n",
       " 'recurrent glioblastoma or other glioma subtype',\n",
       " 'influenza a virus, h7n9 subtype',\n",
       " 'prevention of pneumonia,invasive disease caused by the serotype in 13vpnc',\n",
       " 'other tumor type with fgfr2 gene fusions',\n",
       " 'at10tion-deficit/hyperactivity disorder, predominantly inat10tive type',\n",
       " 'desmoid-type fibromatosis',\n",
       " 'ras-wild-type',\n",
       " 'other b-cell nhl subtype, including wm',\n",
       " 'chronic plaque type psoriasis',\n",
       " 'skin type from i-vi on non-hair bearing skin',\n",
       " 'mixed phenotype acute leukemia, b/myeloid, not otherwise specified',\n",
       " 'mixed phenotype acute leukemia, t/myeloid, not otherwise specified',\n",
       " 'recurrent mixed phenotype acute leukemia',\n",
       " 'chronic 10sion-type headaches',\n",
       " 'ehlers-danlos syndrome hypermobility type',\n",
       " 'at10tion-deficit/hyperactivity disorder (combined-subtype)',\n",
       " \"agitation associated with dementia of the alzheimer's type\",\n",
       " 'niemann-pick c subtype 2',\n",
       " 'adhd, predominantly inat10tive type',\n",
       " 'kras,nras wild-type colorectal cancer',\n",
       " 'cervical spondylosis of cervical type',\n",
       " 'aorta-iliac segment lesion (c,d type by tasc ii)',\n",
       " 'major sickle cell syndrome of type ss or sβ thalassemia',\n",
       " 'mixed phenotype acute leukemia (mpal)',\n",
       " 'ischemic-type biliary lesions',\n",
       " 'extranodal nk/t-cell lymphoma,nasal-type,',\n",
       " 'neovascular age-related macular degeneration of all subtype',\n",
       " 'severe plaque type psoriasis',\n",
       " 'healthy subjects, male or female, of 18-60 years old, with phototype i, ii or iii, not pre-treated/pre-protected skin',\n",
       " 'ras wild type',\n",
       " 'squamous cell carcinoma of the penis, usual type',\n",
       " 'adhd inat10tion or mixed type',\n",
       " 'desmoid type-fibromatosis',\n",
       " 'distal myopathy, nonaka type',\n",
       " 'plaque type psorisis',\n",
       " 'braf wildtype',\n",
       " 'nras wildtype',\n",
       " 'mucopolysaccharidosis type vi',\n",
       " \"agitation in patients with dementia of the alzheimer's type\",\n",
       " 'recurrent diffuse large b-cell lymphoma activated b-cell type',\n",
       " 'refractory diffuse large b-cell lymphoma activated b-cell type',\n",
       " 'chronic severe plaque-type psoriasis',\n",
       " \"dementia of alzheimer's type\",\n",
       " 'extranodal natural killer/t-cell lymphoma, nasal type',\n",
       " 'botryoid-type embryonal rhabdomyosarcoma',\n",
       " 'mixed phenotype acute leukemia in complete remission',\n",
       " 'diffuse large b-cell lymphoma (dlbcl), activated b-cell (abc) subtype',\n",
       " 'non-squamous cell lung cancer with wild-type kras',\n",
       " 'usual type vulval intraepithelial neoplasia (uvin)',\n",
       " 'ehlers-danlos syndrome, vascular type',\n",
       " 'glycogenin-1 deficiency (glycogen storage disease type xv)',\n",
       " 'b precursor type acute leukemia',\n",
       " 'extranodal nk-t-cell lymphoma, nasal type',\n",
       " 'diseases caused by streptococcus pneumoniae serotype',\n",
       " 'enteropathy-type intestinal t-cell lymphoma',\n",
       " 'pancreatic intraductal papillary mucinous neoplasm, pancreatobiliary-type',\n",
       " 'neoplasm by histologic type',\n",
       " 'disease type and/or category not applicable',\n",
       " 'teratoma with somatic-type malignancy',\n",
       " 'thin gingival biotype',\n",
       " 'ampulla of vater intestinal-type adenoma',\n",
       " 'mixed phenotype acute leukemia with t(9;22)(q34.1;q11.2); bcr-abl1',\n",
       " 'refractory mixed phenotype acute leukemia',\n",
       " 'chronic rhinosinusitis phenotype with nasal polyps (crswnp)',\n",
       " 'abnormal karyotype',\n",
       " 'control manifestations of the manic type of mani-depressive illness',\n",
       " 'sleep disorder circadian rhythm, delayed sleep phase type',\n",
       " 't-cell type acute leukemia-precursor t-lymphoblastic lymphoma/leukaemia',\n",
       " 'hypersomnia type; sleep disorder',\n",
       " 'idh family wildtype',\n",
       " 'diffuse large b-cell lymphoma activated b-cell type',\n",
       " 'diffuse large b-cell lymphoma germinal center b-cell type',\n",
       " 'primary cutaneous diffuse large b-cell lymphoma, leg type',\n",
       " 'refractory nasal type extranodal nk/t-cell lymphoma',\n",
       " 'complex karyotype',\n",
       " 'g. breast cancer (triple negative type)',\n",
       " 'liposarcoma; mixed type',\n",
       " 'natural killer/t-cell lymphoma, nasal,nasal-type',\n",
       " 'histological type of neoplasm',\n",
       " 'advanced egfrmutant nonsmallselllungcancer (nsclc),kras g12c nsclc,crc,esophageal squamouscellcancer (scc),head/neck scc,ras/raf wild-type other solid tumor',\n",
       " 'braf v600 wild type',\n",
       " \"alzheimer's disease (incl subtype)\",\n",
       " 'extrahepatic bile duct adenocarcinoma, biliary type',\n",
       " 'gallbladder adenocarcinoma, biliary type',\n",
       " 'ras wild type colorectal cancer',\n",
       " 'advanced brafv600 wild-type melanoma',\n",
       " 'esophageal gastric cardia type metaplasia',\n",
       " 'genotyp_, metabotype,dietary habits',\n",
       " 'jet lag type insomnia',\n",
       " 'medulloblastoma, shh-activated,tp53 wildtype',\n",
       " 'glioblastoma, idh-wildtype',\n",
       " 'shift work type circadian rhythm sleep disorder',\n",
       " 'familial attr-cm (attrm-cm, or fac),wild-type attr-cm (attrwt-cm)',\n",
       " 'circadian rhythm sleep disorder, unspecified type',\n",
       " 'dermatomyositis, adult type',\n",
       " 'cpg island methylator phenotype',\n",
       " 'short sleep phenotype',\n",
       " 'delayed type hypersensitivity',\n",
       " 'dementia; alzheimer, mixed type (etiology)',\n",
       " 'mitochondrial dna depletion syndrome 2 myopathic type',\n",
       " 'negative type; schizophrenic',\n",
       " 'pi*zz, pi*sz, or pi*null phenotype',\n",
       " 'insomnia type; sleep disorder',\n",
       " 'zz type alpha-1 antitrypsin deficiency liver disease',\n",
       " 'nightmares, rem-sleep type',\n",
       " 'muscular dystrophy, duchenne,becker type',\n",
       " 'any type of severe combined immunodeficiency (scid)',\n",
       " 'resistant brca wild-type ovarian cancer',\n",
       " 'schizo-affective type of psychosis',\n",
       " 'c type natriuretic peptide',\n",
       " 'ehlers-danlos syndrome hypermobility type (heds)',\n",
       " 'pancreatobiliary-type carcinoma',\n",
       " \"behcet's syndrome, vascular type\",\n",
       " 'ras,braf wild-type',\n",
       " 'brca wild type platinum sensitive recurrent ovarian cancer',\n",
       " 'classical hodgkin lymphoma type ptld']"
      ]
     },
     "execution_count": 1149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "skipped_dict['type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('2', 5425),\n",
       " ('1', 2119),\n",
       " ('b', 124),\n",
       " ('3', 40),\n",
       " ('a', 22),\n",
       " ('1a', 13),\n",
       " ('c', 12),\n",
       " ('6', 11),\n",
       " ('4', 11),\n",
       " ('2b', 9),\n",
       " ('16', 8),\n",
       " ('2a', 7),\n",
       " ('c1', 7),\n",
       " ('18', 6),\n",
       " ('5', 6),\n",
       " ('11', 4),\n",
       " ('2;', 4),\n",
       " ('7', 3),\n",
       " ('b3', 3),\n",
       " ('d', 2)]"
      ]
     },
     "execution_count": 1150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counters['type'].most_common(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['low grade astrocytomas',\n",
       " 'low-grade astrocytoma, nos',\n",
       " 'low-grade lymphoma',\n",
       " 'high grade glioma',\n",
       " 'lymphoma, low-grade',\n",
       " 'childhood high-grade cerebellar astrocytoma',\n",
       " 'childhood high-grade cerebral astrocytoma',\n",
       " 'childhood low-grade cerebral astrocytoma',\n",
       " 'childhood low-grade cerebellar astrocytoma',\n",
       " 'intermediate-grade lymphoma',\n",
       " 'high-grade squamous intraepithelial lesion',\n",
       " 'low-grade squamous intraepithelial lesion',\n",
       " 'high-grade salivary gland carcinoma',\n",
       " 'high-grade salivary gland mucoepidermoid carcinoma',\n",
       " 'low-grade salivary gland carcinoma',\n",
       " 'low-grade salivary gland mucoepidermoid carcinoma',\n",
       " 'low grade lymphoma',\n",
       " \"low-grade or follicular b-cell non-hodgkin's lymphoma\",\n",
       " 'localised high grade osteosarcoma of the limbs',\n",
       " 'high-grade lymphoma',\n",
       " 'pressure ulcer, grade',\n",
       " 'recurrent high-grade gliomas',\n",
       " 'progressive low-grade gliomas',\n",
       " 'high grade gliomas',\n",
       " 'lymphoma, intermediate-grade',\n",
       " 'lymphoma, high-grade',\n",
       " 'high-grade gliomas',\n",
       " 'myelodysplastic syndrome, high grade',\n",
       " 'high grade squamous intraepithelial neoplasia',\n",
       " 'low grade gliomas',\n",
       " 'non-brainstem high-grade glioma',\n",
       " 'high grade dysplasia',\n",
       " 'low grade dysplasia',\n",
       " 'high-grade glioma',\n",
       " 'low grade ovarian serous adenocarcinoma',\n",
       " 'primary perit1al low grade serous adenocarcinoma',\n",
       " 'malignant lymphomas high grade',\n",
       " 'high-grade glioma of the brain',\n",
       " 'renal s10ting in the treatment of pts w/a high grade ostial atherosclerotic renal lesion(s).',\n",
       " 'adult undifferentiated high grade pleomorphic sarcoma of b1',\n",
       " 'low-grade glioma',\n",
       " 'high-grade myelodysplastic syndrome',\n",
       " 'low grade cns tumors',\n",
       " 'high grade cns tumors',\n",
       " 'low-grade inflammation',\n",
       " 'high-grade lymphomas',\n",
       " 'newly diagnosed high-grade gliomas',\n",
       " 'low-grade gliomas',\n",
       " 'endoscopic retrograde cholangiopancreatography (ercp)',\n",
       " 'grade i/ii ankle sprain',\n",
       " 'high grade sarcoma',\n",
       " 'high-grade precancer',\n",
       " 'high grade cervical intra-epithelial neoplasia',\n",
       " 'endoscopic retrograde cholangiopancreatography',\n",
       " 'high grade astrocytoma',\n",
       " 'low grade glioma',\n",
       " 'low grade dysplasia in ulcerative colitis',\n",
       " \"barrett's mucosa with high grade intraepithelial neoplasia (hgien)\",\n",
       " 'recurrent pediatric high grade glioma',\n",
       " 'recurrent pediatric low grade glioma',\n",
       " 'newly diagnosed pediatric high grade glioma',\n",
       " \"high grade non-hodgkin's lymphoma\",\n",
       " 'recurrent high grade glioma',\n",
       " 'low grade prostate cancer',\n",
       " 'high-grade dysplasia in barrett esophagus',\n",
       " 'high grade myelodysplastic syndrome lesions',\n",
       " 'high-grade meningioma',\n",
       " 'recurrent or progressive low-grade glioma',\n",
       " 'high grade glioma: gliosarcoma',\n",
       " 'high grade glioma: glioblastoma (gbm)',\n",
       " 'progressive or refractory high-grade glioma',\n",
       " 'low-grade chronic inflammation',\n",
       " 'high grade lymphomas',\n",
       " 'high-grade cervical intraepithelial neoplasia',\n",
       " 'low grade cervical squamous intraepithelial neoplasia',\n",
       " 'b-cell malignancy, low-grade',\n",
       " 'low grade metabolic acidosis',\n",
       " 'high grade brain tumor',\n",
       " 'newly diagnosed high-grade glioma',\n",
       " 'esophageal high-grade intraepithelial neoplasia',\n",
       " 'pediatric recurrent progressive low-grade gliomas',\n",
       " 'pediatric progressive low-grade gliomas',\n",
       " 'actinic keratosis olsen grade i/ii',\n",
       " 'cholangiopancreatography, endoscopic retrograde',\n",
       " 'low-grade serous ovarian cancer',\n",
       " 'low-grade serous fallopian tube cancer',\n",
       " 'low-grade serous perit1al cancer',\n",
       " 'high grade serous ovarian cancer',\n",
       " 'high grade intraepithelial neoplasia',\n",
       " 'recurrent, platinum resistant high grade serous ovarian, fallopian tube, or primary perit1al cancer',\n",
       " 'recurrent low-grade gliomas',\n",
       " 'refractory low-grade gliomas',\n",
       " 'bladder cancer transitional cell grade',\n",
       " 'cervical high grade squamous intraepithelial lesion',\n",
       " 'high grade cervical dysplasia',\n",
       " 'high grade lymphomas cd20 positive',\n",
       " 'hiv-negative subjects with intra-anal intraepithelial neoplasia containing high-grade squamous intraepithelial lesions',\n",
       " 'grade approach',\n",
       " 'surgical field grade (condition of surgical field; bleeding, 10sion of the ivc)',\n",
       " 'platinum sensitive recurrent high-grade serous ovarian cancer with mutated p53',\n",
       " 'high grade dysplasia in barrett esophagus',\n",
       " 'recurrent low grade ovarian serous adenocarcinoma',\n",
       " 'post low grade glioma surgery',\n",
       " 'vulvar high grade squamous intraepithelial lesion',\n",
       " 'high grade cervical squamous intraepithelial neoplasia',\n",
       " 'high grade vaginal intraepithelial neoplasia',\n",
       " 'low grade vaginal intraepithelial neoplasia',\n",
       " 'high grade',\n",
       " 'low grade fibromyxoid sarcoma',\n",
       " 'low grade myofibroblastic sarcoma',\n",
       " 'undifferentiated high grade pleomorphic sarcoma of b1',\n",
       " 'high-grade squamous intraepithelial lesions (hsil)',\n",
       " 'high grade ovarian serous adenocarcinoma',\n",
       " 'recurrent high grade meningioma',\n",
       " 'low grade dysplasia oral cavity',\n",
       " 'colorectal high grade intraepithelial neoplasia',\n",
       " 'colorectal low grade intraepithelial neoplasia',\n",
       " 'low grade glioma (lgg), high grade glioma (hgg)',\n",
       " 'newly diagnosed high grade glioma',\n",
       " 'high grade osteosarcoma',\n",
       " 'low grade glioma of brain',\n",
       " 'high-grade serous ovarian cancer',\n",
       " 'lsil, low-grade squamous intraepithelial lesions',\n",
       " 'recurrent high-grade glioma',\n",
       " 'high grade cervical intraepithelial neoplasia',\n",
       " 'low grade glioma (lgg)',\n",
       " 'high grade glioma (hgg)',\n",
       " 'myelodysplastic syndrome with low-grade lesions',\n",
       " \"follicular low grade non-hodgkin's lymphoma\",\n",
       " 'high grade malignant neoplasm',\n",
       " 'high grade glioma (iii or iv)',\n",
       " 'high-grade squamous intraepithelial lesions',\n",
       " 'newly diagnosed high grade glioma (hgg)',\n",
       " 'bithalamic high grade glioma',\n",
       " 'endoscopic retrograde cholangiopancreatogram (ercp)',\n",
       " 'low-grade squamous intraepithelial lesions',\n",
       " 'high grade glioma, glioblastoma',\n",
       " 'high grade fallopian tube serous adenocarcinoma',\n",
       " 'primary perit1al high grade serous adenocarcinoma',\n",
       " 'high grade gastroenteropancreatic neuroendocrine carcinoma',\n",
       " 'pancreatic duct cannulation during endoscopic retrograde choledochopancreatography',\n",
       " 'high-grade upper tract urothelial carcinoma',\n",
       " 'retrograde femoral catheterism',\n",
       " 'top-grade embryo',\n",
       " 'high grade cervical epithelial neoplasia (cin2+)',\n",
       " 'high grade anal canal squamous intraepithelial neoplasia',\n",
       " 'b. high grade serous ovarian cancer',\n",
       " \"non-hodgkin's lymphoma, adult high grade\",\n",
       " 'high grade squamous intra-epithelial lesion (hsil)',\n",
       " 'high grade malignant neuroendocrine carcinoma (diagnosis)',\n",
       " 'vulvar high grade squamous intraepithelial lesion (hsil)',\n",
       " 'high grade squamous intraepithelial lesion',\n",
       " 'hsil, high-grade squamous intraepithelial lesions',\n",
       " 'barretts esophagus with high grade dysplasia',\n",
       " 'high grade endometrial stromal sarcoma',\n",
       " 'high grade serous ovarian',\n",
       " 'face aging/photo aging of moderate grade',\n",
       " 'high grade meningioma',\n",
       " 'mitral valve regurgitation of any grade',\n",
       " 'hsil, high grade squamous intraepithelial lesions',\n",
       " \"high grade non-hodgkin's lymphoma, adult\",\n",
       " 'advanced high-grade ovarian cancer',\n",
       " 'low grade lung neuroendocrine neoplasm',\n",
       " 'intermediate grade lung neuroendocrine neoplasm',\n",
       " 'academic achievement/average grade',\n",
       " 'high-grade astrocytoma nos',\n",
       " 'neuroepithelial tumor, high grade',\n",
       " 'pineal parenchymal tumor of intermediate differentiation (high-grade only)',\n",
       " 'papillary tumor of the pineal region (high-grade only)',\n",
       " 'glioma, high grade',\n",
       " 'glioma, recurrent high grade',\n",
       " 'high grade serous carcinoma',\n",
       " 'high grade recurrent glioma,newly diagnosed glioblastoma',\n",
       " 'high-grade myeloid neoplasm',\n",
       " 'high grade myeloid neoplasm',\n",
       " 'mutant high grade glioma',\n",
       " 'high grade myelodysplastic syndromes',\n",
       " 'endoscopic retrograde cholangiography',\n",
       " 'low grade dysplasia in barrett esophagus',\n",
       " 'low grade endometrial stromal sarcoma',\n",
       " 'low grade serous carcinoma',\n",
       " 'arthroscopic surgical treatment of high-grade (>50%) partialthickness',\n",
       " 'conscious sedation during endoscopic retrograde choledochopancreatography',\n",
       " 'pediatric low-grade gliomas',\n",
       " \"barrett's esophagus with low grade dysplasia\",\n",
       " \"barrett's esophagus with high grade dysplasia\",\n",
       " 'podologic grade from 0 to 3',\n",
       " 'patients undergoing endoscopic retrograde cholangiopancreatography in pr1 position',\n",
       " 'low grade serous ovarian cancer',\n",
       " 'high grade neuroendocrine carcinoma, any site',\n",
       " 'low grade serous ovarian/primary perit1al cancer',\n",
       " 'asc-h atypical squamous cells, cannot rule out high-grade squamous intra-epithelial lesion',\n",
       " 'high-grade extra pulmonary neuroendocrine cancer',\n",
       " 'barretts esophagus with low grade dysplasia',\n",
       " 'high-grade anal intraepithelial neoplasia',\n",
       " 'recurrent low grade fallopian tube serous adenocarcinoma',\n",
       " 'recurrent primary perit1al low grade serous adenocarcinoma',\n",
       " 'high grade serous ovarian carcinoma',\n",
       " 'acute graft-versus-host disease (gvhd) grade',\n",
       " 'students of the 6. grade']"
      ]
     },
     "execution_count": 1153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "skipped_dict['grade']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('3', 537),\n",
       " ('2', 447),\n",
       " ('1', 418),\n",
       " ('b', 43),\n",
       " ('4', 36),\n",
       " ('3a', 17),\n",
       " ('3b', 11),\n",
       " ('d', 2),\n",
       " ('8', 2),\n",
       " ('c', 1)]"
      ]
     },
     "execution_count": 1154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counters['grade'].most_common(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1155,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hepatitis, viral, human',\n",
       " 'nonalcoholic steatohepatitis',\n",
       " 'hepatitis',\n",
       " 'hepatitis, chronic active',\n",
       " 'alcoholic hepatitis',\n",
       " 'hepatitis, alcoholic',\n",
       " 'hepatitis, chronic',\n",
       " 'sequelae of viral hepatitis',\n",
       " 'non alcoholic steatohepatitis',\n",
       " 'chronic hepatitis',\n",
       " 'autoimmune hepatitis',\n",
       " 'non-alcoholic steatohepatitis (nash)',\n",
       " 'non-alcoholic steatohepatitis',\n",
       " 'nash-nonalcoholic steatohepatitis',\n",
       " 'non-alcoholic steatohepatitis(nash)',\n",
       " 'acute ebv hepatitis',\n",
       " 'acute cmv hepatitis',\n",
       " 'acute hepatitis',\n",
       " 'liver cirrhosis due to virus c chronic hepatitis',\n",
       " 'nonalcoholic steatohepatitis (nash)',\n",
       " 'steatohepatitis',\n",
       " 'acute on chronic hepatitis',\n",
       " 'nevirapine induced hepatitis',\n",
       " 'viral hepatitis',\n",
       " 'hepatitis, toxic',\n",
       " 'nash (non-alcoholic steato-hepatitis)',\n",
       " 'chronic alcoholic hepatitis',\n",
       " 'infections, human immunodeficiency virus,hepatitis',\n",
       " 'acute alcoholic hepatitis',\n",
       " 'chronic hcv-related hepatitis',\n",
       " 'ischemic hepatitis',\n",
       " 'hypoxic hepatitis',\n",
       " 'severe alcoholic hepatitis',\n",
       " 'severe acute alcoholic hepatitis',\n",
       " 'work productivity,activity impairment in workers with chronic c hepatitis',\n",
       " 'fatigue in workers with chronic c hepatitis',\n",
       " '•non-alcoholic steatohepatitis (nash)',\n",
       " 'hepatitis, autoimmune',\n",
       " 'non alcoholic steatohepatitis (nash)',\n",
       " \"severe alcoholic hepatitis in 'extremis'-defined by mdf>54\",\n",
       " 'chronic delta hepatitis',\n",
       " 'nash (non-alcoholic steatohepatitis)',\n",
       " 'nonalcoholic steatohepatitis (nash) with fibrosis',\n",
       " 'hepatitis, viral',\n",
       " 'alcoholic hepatitis (ah)',\n",
       " 'hepatitis viruses',\n",
       " 'non-alcoholic steatohepatitis nash',\n",
       " 'chronic hepatitis, c virus',\n",
       " 'hepatitis; virus, chronic, type b',\n",
       " 'carrier of viral hepatitis type b',\n",
       " 'chronic hepatitis, b virus',\n",
       " 'sexually transmitted infections (not hiv or hepatitis)',\n",
       " 'obese patients with non-alcoholic steatohepatitis (nash)',\n",
       " 'nash-nonalcoholic steatohepatitis; nafld-nonalcoholic fatty liver disease',\n",
       " 'severe acute decompensated alcoholic hepatitis',\n",
       " 'chronic hcv hepatitis',\n",
       " 'immune hepatitis',\n",
       " 'steatohepatitis, nonalcoholic']"
      ]
     },
     "execution_count": 1155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "skipped_dict['hepatitis']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('c', 1592), ('b', 883), ('a', 63), ('d', 23), ('e', 16)]"
      ]
     },
     "execution_count": 1156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counters['hepatitis'].most_common(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('v7', 2178), ('v8', 2050), ('v6', 564)]"
      ]
     },
     "execution_count": 1157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counters['ajcc'].most_common(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['type', 'genotyp_', 'grade', 'stage', 'ajcc', 'hepatitis']"
      ]
     },
     "execution_count": 1158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_keywords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['2', '1', '6', '11', '16', '18', 'b', '4', '3', '1a', 'b;', '2a', '2b', 'a', '5', 'c', '2dm', 'a2', 'a3', 'c1', '2c', '0', 'd', '7', '2d', 'b3', 'b2', '3b', '2;', '8', '10', '14', '3a', '2e', '9']\n",
      "['1', '3', '1a', '4', '1b', '5', '2']\n",
      "['1', '2', '3', '4', '3b', 'b', '3a', 'c', 'd', '8']\n",
      "['1', '2', '3', '4a', '4b', '1b', '4', '2b', '2c', '2a', '3b', '0', '3c', '3a', '1a', '4s', '4c', '5', 'a', 'b', 'c', '3c1', '3c2', '0a', '0is', 'b1', '1c', '3a1', '3a1i', '3a1ii', '3a2', '1b2', '1a1', '1a2', '1b1', '5d', 'd', '2a1', '2a2', 'd0', 'her2+', '1a3', '5t', '3d', '4a2', '4a1', '4m1a']\n",
      "['v7', 'v6', 'v8']\n",
      "['b', 'c', 'd', 'a', 'e']\n"
     ]
    }
   ],
   "source": [
    "for cur_c in [raw_counters[x] for x in all_keywords]:\n",
    "    print(list(cur_c.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1041,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hypercholesterolemia with type2dm']"
      ]
     },
     "execution_count": 1041,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig_kw_dict['type']['2dm']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re \n",
    "\n",
    "full_dissim_score = 10.\n",
    "\n",
    "def list_jaccard_dist(l1, l2):\n",
    "    \"\"\" provided 2 lists, compute the jacard similarity between them \"\"\"\n",
    "    full_list = []\n",
    "    for cur_term in l1:\n",
    "        full_list.append(cur_term)\n",
    "    for cur_term in l2:\n",
    "        if cur_term not in full_list:\n",
    "            full_list.append(cur_term)\n",
    "\n",
    "    if len(full_list) == 0:\n",
    "        return -1.\n",
    "    intersect_count = 0\n",
    "    for term in full_list:\n",
    "        if term in l1 and term in l2:\n",
    "            intersect_count += 1\n",
    "\n",
    "    return intersect_count / len(full_list)\n",
    "\n",
    "\n",
    "def digstr2parts(digstr):\n",
    "    \"\"\" returns a tuple of the digit and the string parts, if no num part, returns Null for that part \"\"\"\n",
    "    re_pattern = '^[0-9]+'\n",
    "    f_list = re.findall(re_pattern, digstr)\n",
    "    if len(f_list) == 0:\n",
    "        return None, digstr  # no number part\n",
    "    else:\n",
    "        num_part = int(f_list[0])\n",
    "        chr_part = re.sub(re_pattern, '', digstr)\n",
    "        return num_part, chr_part\n",
    "    \n",
    "    \n",
    "def calc_stage_diff(s1, s2):\n",
    "    \"\"\" compute distance between 3b and 1 for example \"\"\"\n",
    "    enumeratable_letters = 'abcdef'\n",
    "    d1, c1 = digstr2parts(s1)\n",
    "    d2, c2 = digstr2parts(s2)\n",
    "    \n",
    "    # case 1, both digits are None\n",
    "    if d1 is None and d2 is None and \\\n",
    "    c1 in enumeratable_letters and \\\n",
    "    c2 in enumeratable_letters:\n",
    "        return abs(ord(c2) - ord(c1))\n",
    "    \n",
    "    # case 2, 1 digit is None\n",
    "    if d1 is not None and d2 is None:\n",
    "        return full_dissim_score\n",
    "    if d1 is None and d2 is not None:\n",
    "        return full_dissim_score\n",
    "    \n",
    "    # case 3,\n",
    "    if d1 is not None and d2 is not None:\n",
    "        if c1 == '' and c2 == '':\n",
    "            return abs(d2 - d1)\n",
    "        elif c1 != '' and c1 == c2:\n",
    "            return abs(d2 - d1)\n",
    "        else:\n",
    "            return abs(d2 - d1) + .5\n",
    "        \n",
    "    return full_dissim_score\n",
    "        \n",
    "\n",
    "# function to compute simiarity between two extracted stage dicts\n",
    "def stage_sim_dist(stage_dict1, stage_dict2):\n",
    "    \"\"\" stage_dict has key type, grade etc \n",
    "    if dissimilar, returns 10\n",
    "    if stage, type, grade, returns number diff\n",
    "    if both letters, returns char ord diff\n",
    "    \n",
    "    for full_match types applies an (1 - jaccard multiplier) to list diffs\n",
    "    \"\"\"\n",
    "    # if 1 of the dict is blank, return full dissimilarity\n",
    "    if len(stage_dict1) == 0 and len(stage_dict2) == 0:\n",
    "        return 0.\n",
    "    elif len(stage_dict1) == 0 or len(stage_dict2) == 0:\n",
    "        return full_dissim_score\n",
    "    \n",
    "    # hepatitis, genotype, ajcc must be full match\n",
    "    full_sim_required_types = ['hepatitis', 'ajcc', 'genotyp_']\n",
    "    full_match_scores = []\n",
    "    for cur_type in full_sim_required_types:\n",
    "        if len(stage_dict1[cur_type]) > 0 and len(stage_dict2[cur_type]) > 0:\n",
    "            one_m_jdist = 1 - list_jaccard_dist(stage_dict1[cur_type], stage_dict2[cur_type])\n",
    "            full_match_scores.append(one_m_jdist * full_dissim_score)\n",
    "        elif len(stage_dict1[cur_type]) > 0 or len(stage_dict2[cur_type]) > 0:\n",
    "            full_match_scores.append(full_dissim_score)\n",
    "    # if any was filled, then it means some full-match dicts were not null\n",
    "    if len(full_match_scores) > 0:\n",
    "        max_full_score = max(full_match_scores)\n",
    "        if max_full_score > 0:  # only returns if the full match was non-zero other wise check part matches\n",
    "            return max_full_score\n",
    "    \n",
    "    # grade, type, stage can be partially matched\n",
    "    partial_sim_required_types = ['grade', 'type', 'stage']\n",
    "    l1, l2 = [], []\n",
    "    for cur_type in partial_sim_required_types:\n",
    "        for cur_item in stage_dict1[cur_type]:  # append to l1\n",
    "            if cur_item not in l1 and cur_item != '':\n",
    "                l1.append(cur_item)\n",
    "        for cur_item in stage_dict2[cur_type]: # append to l2\n",
    "            if cur_item not in l2 and cur_item != '':\n",
    "                l2.append(cur_item)\n",
    "                \n",
    "    # the elements that are not matching contribute to the difference (if they match, they contribute 0 score)\n",
    "    l1_diff, l2_diff = [], []\n",
    "    l1_diff = [x for x in l1 if x not in l2]\n",
    "    l2_diff = [x for x in l2 if x not in l1]\n",
    "    \n",
    "    # how many were matched?\n",
    "    num_matched = len(l1) - len(l1_diff)\n",
    "    num_diff = len(l1_diff) + len(l2_diff)\n",
    "    \n",
    "    # compute the average distance between l1 and l2 values\n",
    "    all_min_dist = []\n",
    "    for l1_diff_val in l1_diff:\n",
    "        if len(l2_diff) > 0:  # exit condition\n",
    "            all_dist = [calc_stage_diff(l1_diff_val, x) for x in l2_diff]\n",
    "            min_dist = min(all_dist)\n",
    "            min_idx = all_dist.index(min_dist)\n",
    "            del l2_diff[min_idx]\n",
    "            all_min_dist.append(min_dist)\n",
    "        else:\n",
    "            all_min_dist.append(full_dissim_score)\n",
    "    \n",
    "    # if any left in l2_diff, add them as fully dissim\n",
    "    for l2_diff_val in l2_diff:\n",
    "        all_min_dist.append(full_dissim_score)\n",
    "    \n",
    "    if len(all_min_dist) == 0:\n",
    "        return 0.\n",
    "    \n",
    "    mean_dist = sum(all_min_dist) / len(all_min_dist)\n",
    "    return mean_dist * num_diff / (num_matched + num_diff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stage_dict1 = {\n",
    "    'grade':[],\n",
    "    'type':[],\n",
    "    'stage':[],\n",
    "    'genotyp_':[],\n",
    "    'ajcc':['v7'],\n",
    "    'hepatitis':[]\n",
    "}\n",
    "\n",
    "stage_dict2 = {\n",
    "    'grade':['2'],\n",
    "    'type':[],\n",
    "    'stage':[],\n",
    "    'genotyp_':[],\n",
    "    'ajcc':['v7'],\n",
    "    'hepatitis':[]\n",
    "}\n",
    "\n",
    "stage_sim_dist(stage_dict1, stage_dict2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### seen types:\n",
    "- number\n",
    "- roman numerals\n",
    "- english words as numbers\n",
    "- qualifiers that follow the numbers such as a, b, c\n",
    "- \"-\" between valid numbers that signify interval\n",
    "- / that signify and"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
